### PR TITLE
Add logging of headers of DICOMs received by orthanc-raw

### PR DESCRIPTION
**Pull request by jstutters**
Optionally logs selected tags from DICOM instances received by orthanc-raw to a CSV file.

### CODE REVIEW

File: cli/src/pixl_cli/_io.py [hunk 1]
```diff
@@ -67,6 +67,58 @@ def copy_parquet_return_logfile_fields(resources_path: Path) -> tuple[str, datet
...
+        msg = f"csv file expected to have at least {expected_col_names} as " f"column names"
+        raise ValueError(msg)
+
+    (
+        procedure_id_col_name,
+        mrn_col_name,
+        acc_num_col_name,
+        project_col_name,
+        dt_col_name,
+    ) = expected_col_names
```
Comment (stefpiatek) [inline cli/src/pixl_cli/_io.py:96]: should these just be a dictionary of keys and values? where we check that the first 5 values are the `list(dict.values())`

File: cli/src/pixl_cli/_io.py [hunk 2]
```diff
@@ -67,6 +67,58 @@ def copy_parquet_return_logfile_fields(resources_path: Path) -> tuple[str, datet
...
+
+    messages = []
+    for _, row in messages_df.iterrows():
+        message = Message(
+            mrn=row[mrn_col_name],
+            accession_number=row[acc_num_col_name],
+            study_date=datetime.strptime(row[dt_col_name], "%d/%m/%Y %H:%M")
```
Comment (stefpiatek) [inline cli/src/pixl_cli/_io.py:103]: this is the omop es datetime, not the actual datetime

File: cli/src/pixl_cli/_io.py [hunk 3]
```diff
@@ -67,6 +67,58 @@ def copy_parquet_return_logfile_fields(resources_path: Path) -> tuple[str, datet
...
+            .replace(tzinfo=timezone.utc)
+            .date(),
+            procedure_occurrence_id=row[
+                procedure_id_col_name
+            ],  # 4 row[4], #procedure_occurrence_id
+            project_name=row[project_col_name],
+            omop_es_timestamp=omop_es_timestamp,
```
Comment (stefpiatek) [inline cli/src/pixl_cli/_io.py:110]: we should rename the Message field from `omop_es_timestamp` to something more generic like `extract_generated_timestamp`

File: cli/src/pixl_cli/_io.py [hunk 4]
```diff
@@ -122,6 +174,56 @@ def messages_from_parquet(
...
+    timestamp: datetime,
+) -> list[Message]:
+    """
+    Reads patient information from a parquet file and transforms that into messages.
+
+    :param file_path: Path for parquet file
+    """
+    cohort_data = pd.read_parquet(file_path)
+
+    expected_col_names = [
```
Comment (stefpiatek) [inline cli/src/pixl_cli/_io.py:189]: same here for the dict

File: cli/src/pixl_cli/_io.py [hunk 5]
```diff
@@ -67,6 +67,58 @@ def copy_parquet_return_logfile_fields(resources_path: Path) -> tuple[str, datet
...
+        "mrn",
+        "accession_number",
+        "project_name",
+        "omop-es-datetime",
+    ]
+
+    # First line is column names
+    messages_df = pd.read_csv(filepath, header=0, dtype=str)
+
+    if list(messages_df.columns)[:5] != expected_col_names:
```
Comment (stefpiatek) [inline cli/src/pixl_cli/_io.py:86]: could this use the pre-existing `_raise_if_column_names_not_found`?

File: cli/src/pixl_cli/main.py [hunk 6]
```diff
@@ -79,9 +81,13 @@ def populate(parquet_dir: Path, *, restart: bool, queues: str) -> None:
             │   └── PROCEDURE_OCCURRENCE.parquet
             └── extract_summary.json
     """
-    logger.info(f"Populating queue(s) {queues} from {parquet_dir}")
-    project_name, omop_es_datetime = copy_parquet_return_logfile_fields(parquet_dir)
-    messages = messages_from_parquet(parquet_dir, project_name, omop_es_datetime)
+    logger.info(f"Populating queue(s) {queues} from {parquet_path}")
+    if parquet_path.is_file() and parquet_path.suffix == ".csv":
+        omop_es_datetime = datetime.now(tz=timezone.utc)
```
Comment (stefpiatek) [inline cli/src/pixl_cli/main.py:86]: We should name this variable what it actually is

File: cli/tests/resources/test.csv [hunk 7]
```diff
@@ -0,0 +1,2 @@
+procedure_id,mrn,accession_number,project_name,omop-es-datetime
```
Comment (stefpiatek) [inline cli/tests/resources/test.csv:1]: Do we want to require the real study date as well? Only for ordering, but that might be reasonable?

File: docker/orthanc-raw/Dockerfile [hunk 8]
```diff
@@ -16,7 +16,26 @@ SHELL ["/bin/bash", "-o", "pipefail", "-e", "-u", "-x", "-c"]
 
 ARG ORTHANC_RAW_MAXIMUM_STORAGE_SIZE
 
+RUN --mount=type=cache,target=/root/.cache \
+    pip3 install pydicom pyyaml
```
Comment (stefpiatek) [inline docker/orthanc-raw/Dockerfile:20]: Can we make this a pinned dependency in dcmd

File: orthanc/orthanc-raw/plugin/pixl.py [hunk 9]
```diff
@@ -44,6 +49,21 @@ def OnHeartBeat(output, uri, **request):  # noqa: ARG001
...
 
 
+def ReceivedInstanceCallback(receivedDicom: bytes, _: str) -> Any:
+    """Optionally record headers from the received DICOM instance."""
+    if ShouldRecordHeaders():
+        record_dicom_headers(receivedDicom)
+    return orthanc.ReceivedInstanceAction.KEEP_AS_IS, None
+
+
+def ShouldRecordHeaders() -> bool:
```
Comment (stefpiatek) [inline orthanc/orthanc-raw/plugin/pixl.py:59]: Might be worth keeping to snake case for everything that isn't an orthanc callback

File: pixl_dcmd/MANIFEST.in [hunk 10]
```diff
@@ -0,0 +1 @@
+include src/pixl_dcmd/resources/*.yml
```
Comment (stefpiatek) [inline pixl_dcmd/MANIFEST.in:1]: I thin you can do this in pyproject.toml which IIRC we've done in pytest-pixl using package-data

File: pixl_dcmd/src/pixl_dcmd/tagrecording.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+import yaml
+
+
+def record_dicom_headers(receivedDicom: bytes) -> None:
+    with importlib.resources.files("pixl_dcmd").joinpath(
+        "resources/recorded-headers.yml"
+    ).open() as f:
+        recording_config = yaml.safe_load(f)
+    dataset = dcmread(BytesIO(receivedDicom), force=True)
```
Comment (stefpiatek) [inline pixl_dcmd/src/pixl_dcmd/tagrecording.py:31]: why force?

File: docker-compose.yml [hunk 12]
```diff
@@ -182,6 +182,8 @@ services:
       ORTHANC_RAW_AE_TITLE: ${ORTHANC_RAW_AE_TITLE}
       ORTHANC_AUTOROUTE_RAW_TO_ANON: ${ORTHANC_AUTOROUTE_RAW_TO_ANON}
       ORTHANC_RAW_MAXIMUM_STORAGE_SIZE: ${ORTHANC_RAW_MAXIMUM_STORAGE_SIZE}
+      ORTHANC_RAW_RECORD_HEADERS: ${ORTHANC_RAW_RECORD_HEADERS}
+      ORTHANC_RAW_HEADER_LOG_PATH: ${ORTHANC_RAW_HEADER_LOG_PATH}
```
Comment (stefpiatek) [inline docker-compose.yml:186]: can we add these to the sample.env as well 🙏🏻

File: test/scripts/run_cli.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from pixl_cli.main import cli
+
+cli()
```
Comment (stefpiatek) [inline test/scripts/run_cli.py:19]: is this file ever used?

**Review by stefpiatek — APPROVED** [review]
Looks like it should work, I think we can tidy it up. Happy for you to merge when you've addressed the discussions

File: pixl_dcmd/MANIFEST.in (hunk 10 above)
Comment (jstutters) [reply pixl_dcmd/MANIFEST.in:1]: Indeed you can, I thought I'd tried it that way but I must have mucked up the syntax.

File: pixl_dcmd/src/pixl_dcmd/tagrecording.py (hunk 11 above)
Comment (jstutters) [reply pixl_dcmd/src/pixl_dcmd/tagrecording.py:31]: The DICOMs I have to test with have janky headers that require `force` but shouldn't be necessary in prod - removed.

File: test/scripts/run_cli.py (hunk 13 above)
Comment (jstutters) [reply test/scripts/run_cli.py:19]: Nope - for testing/debugging only.

File: cli/src/pixl_cli/_io.py (hunk 1 above)
Comment (jstutters) [reply cli/src/pixl_cli/_io.py]: Managed to trim this down a bit.

**Comment by ruaridhg** [comment]
I have added a fresh commit here for modifying the test.csv, but it's not appearing

**Pull request merged**